lnst/Recipes/ENRT/ConfigMixins/FirewallMixin.py

```diff
@@ -0,0 +1,197 @@
+          self.matched.host2: <host2 ruleset> }
+
+        The rulesets will be applied by a derived class's _apply_ruleset()
+        method, i.e. typically fed into 'nft -f' or 'iptables-restore'.
+        """
+        return {}
+
+    def _apply_ruleset(self, host, path):
+        raise NotImplementedError("Method _apply_ruleset must be defined by a child class.")
+
+    def _dump_ruleset(self, host, config):
+        raise NotImplementedError("Method _dump_ruleset must be defined by a child class.")
+
+    def _push_ruleset(self, host, ruleset):
+        """
+        Push `ruleset` into a temporary file on `host`, returning its path.
+        """
+        fd, name = mkstemp(prefix="FirewallMixin", text=True)
+        write(fd, ruleset.encode('utf-8'))
+        close(fd)
+        remote_path = host.copy_file_to_machine(name)
+        unlink(name)
+        return remote_path
+
+    def apply_sub_configuration(self, config):
+        super().apply_sub_configuration(config)
+
+        if not config.firewall_rulesets:
+            config.firewall_rulesets = self.firewall_rulesets
+
+        self.stored_rulesets = {}
+        for host, ruleset in config.firewall_rulesets.items():
+            self.stored_rulesets[host] = self._dump_ruleset(host, config)
+
+            name = self._push_ruleset(host, ruleset)
+            self._apply_ruleset(host, name)
+            host.run('rm -f "%s"' % name)
+
+    def generate_sub_configuration_description(self, config):
+        desc = super().generate_sub_configuration_description(config)
+
+        for host, ruleset in config.firewall_rulesets.items():
+            desc.append("Firewall: ruleset with %d lines on host %s" % (len(ruleset.split("\n")), host))
+
+        return desc
+
+    def remove_sub_configuration(self, config):
+        for host, ruleset in self.stored_rulesets.items():
+            # XXX: this may spam debug log, but counter values may be worth logging
+            logging.debug("Dropping Firewall ruleset on host %s:\n%s\n" % (host, self._dump_ruleset(host, config)))
+            name = self._push_ruleset(host, ruleset)
+            self._apply_ruleset(host, name)
+            host.run('rm -f "%s"' % name)
+
+        return super().remove_sub_configuration(config)
+
+class BulkFirewallMixin(FirewallMixin):
+    """
+    Extend FirewallMixin by the ability to create hosts' firewall configs using
+    a generator property. The generator may yield multiple times to test
+    different ruleset sizes.
+    """
+
+    @property
+    def bulk_firewall_rules_generator(self):
+        """
+        A generator yielding { host: ruleset, ... } to apply for a test run.
+        Each yield will turn into a new sub configuration and thus be tested
+        separately.
+        """
+        raise NotImplementedError("Method bulk_firewall_rules_generator must be defined by a child class.")
+
+    def generate_sub_configurations(self, config):
+        for ruleset in self.bulk_firewall_rules_generator:
```

Comment:
Could you use plural for "ruleset" since we're assigning it later to `new_config.firewall_rulesets = ruleset`?

Minor thing, but I had to stop for a while why it's just ONE ruleset, so probably worth to change to make it more readable.

Reply:
DONE.